### PR TITLE
Develop: Which Aligner Are you?

### PR DESCRIPTION
**Pull request by J35P312**
There are many aligners! All with different properties, but which one represents you the most?
Begin the test by clicking the Merge button!

********
I made a test and module for GATK splitNcigarreads. This GATK tools split read containing N cigar operation into multiple alignment, also it changes the mapping qualities from 255 to 60.
This software should be run on the bam file before running variant calling and ASE analysis.

### CODE REVIEW

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 1]
```diff
@@ -27,7 +27,7 @@ BEGIN {
 
     # Functions and variables which can be optionally exported
     our @EXPORT_OK =
-      qw{ gatk_realignertargetcreator gatk_indelrealigner gatk_baserecalibrator gatk_printreads gatk_haplotypecaller };
+      qw{ gatk_realignertargetcreator gatk_indelrealigner gatk_baserecalibrator gatk_printreads gatk_haplotypecaller gatk_splitncigarreads };
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Let's try to clean-up while we are modifying code, 😄 . Order this AB.

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 2]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+##          : $outfile_path         => Outfile path
+##          : $operation            => The read operation that is to be applied on the bam file
+##          : $readfilter           => Read-filter settings
+##          : $referencefile_path   => Reference sequence file
+##          : $stderrfile_path      => Stderrfile path
+##          : $temp_directory       => Redirect tmp files to java temp
+
+    my ($arg_href) = @_;
+
+    ## Default(s)
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Move defaults to under ## Flatten argument(s) so it complies with the template

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 3]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+    my $java_jar;
+    my $java_use_large_pages;
+    my $memory_allocation;
+    my $outfile_path;
+    my $stderrfile_path;
+    my $referencefile_path;
+    my $temp_directory;
+
+    my $tmpl = {
+        FILEHANDLE  => { store => \$FILEHANDLE },
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: \$FILEHANDLE -> \$FILEHANDLE,

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 4]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+        infile_path => {
+            required    => 1,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Order AB

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 5]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+            defined     => 1,
+            strict_type => 1,
+            store       => \$infile_path,
+        },
+        java_jar             => { strict_type => 1, store => \$java_jar },
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Always add comma after last element

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 6]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+            store       => \$java_use_large_pages,
+        },
+        logging_level => {
+            default     => q{INFO},
+            allow       => [qw{ INFO ERROR FATAL }],
+            strict_type => 1,
+            store       => \$logging_level,
+        },
+        maping_quality_from => {
+            default     => 255,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Use readonly constants

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 7]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+            defined     => 1,
+            strict_type => 1,
+            store       => \$maping_quality_from,
+        },
+        maping_quality_to => {
+            default     => 60,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Use readonly constants

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 8]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+
+    # Stores commands depending on input parameters
+    my @commands;
+
+    ## Write java core commands to filehandle.
+    if ($java_jar) {
+
+        @commands = java_core(
+            {
+                memory_allocation    => $memory_allocation,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Order AB

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 9]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+                temp_directory       => $temp_directory,
+                java_jar             => $java_jar,
+            }
+        );
+    }
+
+    ### Gatk base args
+    @commands = gatk_base(
+        {
+            commands_ref       => \@commands,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Order AB

File: lib/MIP/Program/Alignment/Gatk.pm [hunk 10]
```diff
@@ -1102,4 +1102,162 @@ sub gatk_haplotypecaller {
...
+      unix_standard_streams(
+        {
+            stderrfile_path => $stderrfile_path,
+        }
+      );
+
+    unix_write_to_file(
+        {
+            commands_ref => \@commands,
+            separator    => $SPACE,
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Alignment/Gatk.pm]: Order AB

File: t/gatk_splitncigarreads.t [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    infile_path => {
+        input           => catfile(qw{ dir infile.bam }),
+        expected_output => q{-I} . $SPACE . catfile(qw{ dir infile.bam }),
+    },
+    outfile_path => {
+        input           => catfile(qw{ dir outfile.bam }),
+        expected_output => q{-o} . $SPACE . catfile(qw{ dir infile.bam }),
+    },
+    referencefile_path => {
+        input           => catfile(qw{reference_dir human_genome_build.fasta }),
```
Comment (henrikstranneheim) [inline t/gatk_splitncigarreads.t]: {r -> { r

File: t/gatk_splitncigarreads.t [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        expected_output => q{--reference_sequence}
+          . $SPACE
+          . catfile(qw{reference_dir human_genome_build.fasta }),
```
Comment (henrikstranneheim) [inline t/gatk_splitncigarreads.t]: {r -> { r

File: t/gatk_splitncigarreads.t [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+    },
+);
+
+my %specific_argument = (
+    infile_path => {
+        input           => catfile(qw{ dir infile.bam }),
+        expected_output => q{-I} . $SPACE . catfile(qw{ dir infile.bam }),
+    },
+    maping_quality_from => {
+        input           => q{60},
```
Comment (henrikstranneheim) [inline t/gatk_splitncigarreads.t]: q{60} is a number (60) and not a string. Use Readonly constants according to perl best practise

File: t/gatk_splitncigarreads.t [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+    outfile_path => {
+        input           => catfile(qw{ dir outfile.bam }),
+        expected_output => q{-o} . $SPACE . catfile(qw{ dir outfile.bam }),
+    },
+    operation => {
+        input           => q{STUFF},
+        expected_output => q{-U} . $SPACE . q{STUFF},
+    },
+    referencefile_path => {
+        input           => catfile(qw{reference_dir human_genome_build.fasta }),
```
Comment (henrikstranneheim) [inline t/gatk_splitncigarreads.t]: {r -> { r

File: t/gatk_splitncigarreads.t [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+        expected_output => q{--reference_sequence}
+          . $SPACE
+          . catfile(qw{reference_dir human_genome_build.fasta }),
```
Comment (henrikstranneheim) [inline t/gatk_splitncigarreads.t]: {r -> { r

**Comment by henrikstranneheim** [comment]
@J35P312 Any progress with this one?

**Comment by J35P312** [comment]
Hello again!
Sorry for the delay, now I have cleaned up the code =P

**Comment by henrikstranneheim** [comment]
Look kind a bit strange, when the modules in the GATK modules were moved around, but I think all was fine and it is just a github representation of code changes.
👍

**Pull request merged**